### PR TITLE
Introduce union-logr(klog, nats-logr)

### PR DESCRIPTION
**Pull request by masudur-rahman**

### CODE REVIEW

File: apiserver/create.go [hunk 1]
```diff
@@ -37,28 +38,31 @@ func (a *Apiserver) Init(storeProvider store.Interface, msg *stan.Msg) (*api.Ope
...
 }
 
 func (a *Apiserver) CreateCluster(storeProvider store.Interface) error {
 	_, err := a.natsConn.QueueSubscribe("create-cluster", "cluster-api-create-workers", func(msg *stan.Msg) {
-		log := klogr.New().WithName("[apiserver]")
+		operation, scope, err := a.Init(storeProvider, msg)
 
-		log.Info("create operation")
+		initLog := natslogr.New().
+			WithValues(natslogr.ClusterID, "pharmer-cluster", natslogr.ClientID, operation.Code, natslogr.NatsURL, stan.DefaultNatsURL, natslogr.ConnectWait, 5).
```
Comment (tahsinrahman) [inline apiserver/create.go]:
```suggestion
			WithValues(natslogr.ClusterID, "pharmer-cluster",
			natslogr.ClientID, operation.Code,
			natslogr.NatsURL, stan.DefaultNatsURL,
			natslogr.ConnectWait, 5).
```

File: apiserver/retry.go [hunk 2]
```diff
@@ -1,26 +1,31 @@
...
 )
 
 func (a *Apiserver) RetryCluster(storeProvider store.Interface) error {
 	_, err := a.natsConn.QueueSubscribe("retry-cluster", "cluster-api-retry-workers", func(msg *stan.Msg) {
-		log := klogr.New().WithName("[apiserver]")
+		operation, scope, err := a.Init(storeProvider, msg)
 
-		log.Info("retry operation")
+		initLog := natslogr.New().
+			WithValues(natslogr.ClusterID, "pharmer-cluster", natslogr.ClientID, operation.Code, natslogr.NatsURL, stan.DefaultNatsURL, natslogr.ConnectWait, 5).
```
Comment (tahsinrahman) [inline apiserver/retry.go]:
```suggestion
			WithValues(natslogr.ClusterID, "pharmer-cluster",
			natslogr.ClientID, operation.Code,
			natslogr.NatsURL, stan.DefaultNatsURL,
			natslogr.ConnectWait, 5).
```

File: go.mod [hunk 3]
```diff
@@ -40,14 +36,7 @@ require (
...
 	github.com/mitchellh/go-wordwrap v1.0.0 // indirect
-	github.com/nats-io/gnatsd v1.4.1 // indirect
-	github.com/nats-io/go-nats v1.7.2 // indirect
-	github.com/nats-io/go-nats-streaming v0.4.4 // indirect
-	github.com/nats-io/nats-server v1.4.1 // indirect
-	github.com/nats-io/nats-streaming-server v0.14.1 // indirect
-	github.com/nats-io/nkeys v0.0.1 // indirect
-	github.com/nats-io/nuid v1.0.1 // indirect
-	github.com/nats-io/stan.go v0.4.4
+	github.com/nats-io/go-nats-streaming v0.4.4
```
Comment (tahsinrahman) [inline go.mod]: can you check this?
```
This repo exists to support projects that used this NATS streaming client prior to the name change and go.mod support.

See nats-io/stan.go instead.
```

File: go.mod [hunk 4]
```diff
@@ -64,6 +53,7 @@ require (
 	golang.org/x/oauth2 v0.0.0-20190402181905-9f3314589c9a
 	golang.org/x/sys v0.0.0-20190613124609-5ed2794edfdc // indirect
 	gomodules.xyz/cert v1.0.0
+	gomodules.xyz/nats-logr v0.0.0-20190702081800-7457beea95bc
```
Comment (tahsinrahman) [inline go.mod]: this should be changed to master/release after `gomodules.xyz/nats-logr` is released

File: apiserver/create.go [hunk 5]
```diff
@@ -2,14 +2,15 @@ package apiserver
 
 import (
 	"encoding/json"
+	"errors"
 	"strconv"
 
-	"github.com/nats-io/stan.go"
+	stan "github.com/nats-io/go-nats-streaming"
```
Comment (tahsinrahman) [inline apiserver/create.go]: this repo is archived. should we use this?

File: apiserver/delete.go [hunk 6]
```diff
@@ -1,26 +1,31 @@
 package apiserver
 
 import (
-	"github.com/nats-io/stan.go"
+	stan "github.com/nats-io/go-nats-streaming"
```
Comment (tahsinrahman) [inline apiserver/delete.go]: this repo is archived. should we use this?

File: apiserver/nats.go [hunk 7]
```diff
@@ -4,7 +4,7 @@ import (
 	"io"
 	"log"
 
-	"github.com/nats-io/stan.go"
+	stan "github.com/nats-io/go-nats-streaming"
```
Comment (tahsinrahman) [inline apiserver/nats.go]: this repo is archived. should we use this?

File: apiserver/retry.go [hunk 8]
```diff
@@ -1,26 +1,31 @@
 package apiserver
 
 import (
-	"github.com/nats-io/stan.go"
+	stan "github.com/nats-io/go-nats-streaming"
```
Comment (tahsinrahman) [inline apiserver/retry.go]: this repo is archived. should we use this?

File: cmds/server.go [hunk 9]
```diff
@@ -1,10 +1,13 @@
...
 
 import (
+	"flag"
 	"fmt"
 
+	natslog "gomodules.xyz/nats-logr/nats-log"
+
 	"github.com/appscode/go/term"
-	"github.com/nats-io/stan.go"
+	stan "github.com/nats-io/go-nats-streaming"
```
Comment (tahsinrahman) [inline cmds/server.go]: this repo is archived. should we use this?

File: notification/notification.go [hunk 10]
```diff
@@ -4,7 +4,7 @@ import (
 	"fmt"
 
 	"github.com/appscode/go/log"
-	"github.com/nats-io/stan.go"
+	stan "github.com/nats-io/go-nats-streaming"
```
Comment (tahsinrahman) [inline notification/notification.go]: this repo is archived. should we use this?

File: apiserver/create.go [hunk 11]
```diff
@@ -37,28 +39,37 @@ func (a *Apiserver) Init(storeProvider store.Interface, msg *stan.Msg) (*api.Ope
...
+			ClientID:  operation.Code,
+			NatsURL:   stan.DefaultNatsURL,
+			Subject:   scope.Cluster.Name + "-" + strconv.FormatInt(operation.UserID, 10),
+		}
 
-		log.V(4).Info("nats message", "sequence", msg.Sequence, "redelivered", msg.Redelivered,
-			"message string", string(msg.Data))
+		logN := natslogr.NewLogger(opts)
+		ulog := ulogr.NewLogger(logN)
+
```
Comment (tahsinrahman) [inline apiserver/create.go]:
```suggestion
```

File: apiserver/create.go [hunk 12]
```diff
@@ -37,28 +39,37 @@ func (a *Apiserver) Init(storeProvider store.Interface, msg *stan.Msg) (*api.Ope
...
+			ClusterID: "pharmer-cluster",
+			ClientID:  operation.Code,
+			NatsURL:   stan.DefaultNatsURL,
+			Subject:   scope.Cluster.Name + "-" + strconv.FormatInt(operation.UserID, 10),
+		}
 
-		log.V(4).Info("nats message", "sequence", msg.Sequence, "redelivered", msg.Redelivered,
-			"message string", string(msg.Data))
+		logN := natslogr.NewLogger(opts)
+		ulog := ulogr.NewLogger(logN)
```
Comment (tahsinrahman) [inline apiserver/create.go]: why ulogr?

File: apiserver/create.go [hunk 13]
```diff
@@ -37,28 +39,37 @@ func (a *Apiserver) Init(storeProvider store.Interface, msg *stan.Msg) (*api.Ope
...
 	return obj, scope, nil
 }
 
 func (a *Apiserver) CreateCluster(storeProvider store.Interface) error {
 	_, err := a.natsConn.QueueSubscribe("create-cluster", "cluster-api-create-workers", func(msg *stan.Msg) {
-		log := klogr.New().WithName("[apiserver]")
+		operation, scope, err := a.Init(storeProvider, msg)
 
-		log.Info("create operation")
+		opts := natslogr.Options{
```
Comment (tahsinrahman) [inline apiserver/create.go]: common code, move to function?

File: apiserver/delete.go [hunk 14]
```diff
@@ -1,26 +1,40 @@
...
 
 func (a *Apiserver) DeleteCluster(storeProvider store.Interface) error {
 	_, err := a.natsConn.QueueSubscribe("delete-cluster", "cluster-api-delete-workers", func(msg *stan.Msg) {
-		log := klogr.New().WithName("[apiserver]")
+		operation, scope, err := a.Init(storeProvider, msg)
 
-		log.Info("delete operation")
-		log.V(4).Info("nats message", "sequence", msg.Sequence, "redelivered", msg.Redelivered,
-			"message string", string(msg.Data))
+		opts := natslogr.Options{
```
Comment (tahsinrahman) [inline apiserver/delete.go]: common code, move to function?

File: apiserver/nats.go [hunk 15]
```diff
@@ -4,7 +4,7 @@ import (
 	"io"
 	"log"
 
-	"github.com/nats-io/stan.go"
+	stan "github.com/nats-io/stan.go"
```
Comment (tahsinrahman) [inline apiserver/nats.go]:
```suggestion
	"github.com/nats-io/stan.go"
```

File: apiserver/delete.go [hunk 16]
```diff
@@ -1,26 +1,40 @@
 package apiserver
 
 import (
-	"github.com/nats-io/stan.go"
+	"strconv"
+
+	stan "github.com/nats-io/stan.go"
```
Comment (tahsinrahman) [inline apiserver/delete.go]:
```suggestion
	"github.com/nats-io/stan.go"
```

File: apiserver/retry.go [hunk 17]
```diff
@@ -1,26 +1,40 @@
 package apiserver
 
 import (
-	"github.com/nats-io/stan.go"
+	"strconv"
+
+	stan "github.com/nats-io/stan.go"
```
Comment (tahsinrahman) [inline apiserver/retry.go]:
```suggestion
	"github.com/nats-io/stan.go"
```

File: apiserver/retry.go [hunk 18]
```diff
@@ -1,26 +1,40 @@
...
+	ulogr "gomodules.xyz/union-logr"
 )
 
 func (a *Apiserver) RetryCluster(storeProvider store.Interface) error {
 	_, err := a.natsConn.QueueSubscribe("retry-cluster", "cluster-api-retry-workers", func(msg *stan.Msg) {
-		log := klogr.New().WithName("[apiserver]")
+		operation, scope, err := a.Init(storeProvider, msg)
 
-		log.Info("retry operation")
+		opts := natslogr.Options{
```
Comment (tahsinrahman) [inline apiserver/retry.go]: common code, move to function?

File: cloud/providers/digitalocean/cluster_actuator.go [hunk 19]
```diff
@@ -39,11 +40,12 @@ type ClusterActuatorParams struct {
 
 func NewClusterActuator(m manager.Manager, params ClusterActuatorParams) *ClusterActuator {
 	params.cm.Logger = params.cm.Logger.WithName("[cluster-actuator]")
+	logK := klogr.New()
```
Comment (tahsinrahman) [inline cloud/providers/digitalocean/cluster_actuator.go]:
```suggestion
```

File: cloud/providers/digitalocean/cluster_actuator.go [hunk 20]
```diff
@@ -39,11 +40,12 @@ type ClusterActuatorParams struct {
...
 	return &ClusterActuator{
 		client:        m.GetClient(),
 		eventRecorder: params.EventRecorder,
 		cm:            params.cm,
-		Logger: klogr.New().WithName("[cluster-actuator]").
+		Logger: ulogr.NewLogger(logK).WithName("[cluster-actuator]").
```
Comment (tahsinrahman) [inline cloud/providers/digitalocean/cluster_actuator.go]: revert it, it is machine controller pod log

File: cloud/providers/digitalocean/machine_actuator.go [hunk 21]
```diff
@@ -58,12 +59,13 @@ type MachineActuatorParams struct {
 
 func NewMachineActuator(params MachineActuatorParams) *MachineActuator {
 	params.cm.Logger = params.cm.Logger.WithName("[machine-actuator]")
+	logK := klogr.New()
```
Comment (tahsinrahman) [inline cloud/providers/digitalocean/machine_actuator.go]: revert

File: cloud/providers/digitalocean/machine_actuator.go [hunk 22]
```diff
@@ -58,12 +59,13 @@ type MachineActuatorParams struct {
...
 	return &MachineActuator{
 		client:        params.Client,
 		kubeadm:       getKubeadm(params),
 		eventRecorder: params.EventRecorder,
 		cm:            params.cm,
-		Logger: klogr.New().WithName("[machine-actuator]").
+		Logger: ulogr.NewLogger(logK).WithName("[machine-actuator]").
```
Comment (tahsinrahman) [inline cloud/providers/digitalocean/machine_actuator.go]: revert

File: cloud/providers/linode/cluster_actuator.go [hunk 23]
```diff
@@ -42,12 +43,13 @@ type ClusterActuatorParams struct {
 }
 
 func NewClusterActuator(m manager.Manager, params ClusterActuatorParams) *ClusterActuator {
+	logK := klogr.New()
```
Comment (tahsinrahman) [inline cloud/providers/linode/cluster_actuator.go]: revert

File: cloud/providers/linode/cluster_actuator.go [hunk 24]
```diff
@@ -42,12 +43,13 @@ type ClusterActuatorParams struct {
...
 	return &ClusterActuator{
 		client:        m.GetClient(),
 		eventRecorder: params.EventRecorder,
 		scheme:        params.Scheme,
 		cm:            params.cm,
-		Logger: klogr.New().WithName("[cluster-actuator]").
+		Logger: ulogr.NewLogger(logK).WithName("[cluster-actuator]").
```
Comment (tahsinrahman) [inline cloud/providers/linode/cluster_actuator.go]: revert

File: cloud/providers/linode/machine_actuator.go [hunk 25]
```diff
@@ -53,12 +54,13 @@ type MachineActuatorParams struct {
 }
 
 func NewMachineActuator(params MachineActuatorParams) *MachineActuator {
+	logK := klogr.New()
```
Comment (tahsinrahman) [inline cloud/providers/linode/machine_actuator.go]: revert

File: cloud/providers/linode/machine_actuator.go [hunk 26]
```diff
@@ -53,12 +54,13 @@ type MachineActuatorParams struct {
...
 	return &MachineActuator{
 		cm:            params.cm,
 		client:        params.Client,
 		kubeadm:       getKubeadm(params),
 		eventRecorder: params.EventRecorder,
-		Logger: klogr.New().WithName("[machine-actuator]").
+		Logger: ulogr.NewLogger(logK).WithName("[machine-actuator]").
```
Comment (tahsinrahman) [inline cloud/providers/linode/machine_actuator.go]: revert

File: cloud/providers/packet/cluster_actuator.go [hunk 27]
```diff
@@ -36,7 +37,8 @@ type ClusterActuatorParams struct {
 }
 
 func NewClusterActuator(m manager.Manager, params ClusterActuatorParams) *ClusterActuator {
-	params.cm.Logger = klogr.New().WithName("[cachine-actuator]").
+	logK := klogr.New()
```
Comment (tahsinrahman) [inline cloud/providers/packet/cluster_actuator.go]: revert

File: cloud/providers/packet/machine_actuator.go [hunk 28]
```diff
@@ -52,7 +53,8 @@ type MachineActuatorParams struct {
 }
 
 func NewMachineActuator(params MachineActuatorParams) *MachineActuator {
-	params.cm.Logger = klogr.New().WithName("[machine-actuator]").
+	logK := klogr.New()
```
Comment (tahsinrahman) [inline cloud/providers/packet/machine_actuator.go]: revert

File: cloud/scope.go [hunk 29]
```diff
@@ -42,7 +43,8 @@ type NewScopeParams struct {
 
 func NewScope(params NewScopeParams) *Scope {
 	if params.Logger == nil {
-		params.Logger = klogr.New().WithValues("cluster-name", params.Cluster.Name)
+		logK := klogr.New()
```
Comment (tahsinrahman) [inline cloud/scope.go]: Q: is it needed here? scope.logger is set from `apiserver/`

File: cmds/cloud/apply.go [hunk 30]
```diff
@@ -50,11 +51,12 @@ func runApplyCmd(storeProvider store.ResourceInterface, opts *options.ApplyConfi
 		return errors.Wrapf(err, "failed to get certs")
 	}
 
+	logK := klogr.New()
```
Comment (tahsinrahman) [inline cmds/cloud/apply.go]: Q: is it needed in cli cmd?

File: cmds/cloud/create_cluster.go [hunk 31]
```diff
@@ -43,10 +44,11 @@ func NewCmdCreateCluster() *cobra.Command {
 }
 
 func runCreateClusterCmd(store store.ResourceInterface, opts *options.ClusterCreateConfig) error {
+	logK := klogr.New()
```
Comment (tahsinrahman) [inline cmds/cloud/create_cluster.go]: Q: is it needed in cli cmd?

File: cmds/controller.go [hunk 32]
```diff
@@ -36,10 +37,11 @@ func newCmdController() *cobra.Command {
 				term.Fatalln(err)
 			}
 
+			logK := klogr.New()
```
Comment (tahsinrahman) [inline cmds/controller.go]: revert. it is not needed in machine controller pod

File: cmds/server.go [hunk 33]
```diff
@@ -1,10 +1,13 @@
...
 
 import (
+	"flag"
 	"fmt"
 
+	natslogr "gomodules.xyz/nats-logr"
+
 	"github.com/appscode/go/term"
-	"github.com/nats-io/stan.go"
+	stan "github.com/nats-io/stan.go"
```
Comment (tahsinrahman) [inline cmds/server.go]:
```suggestion
	"github.com/nats-io/stan.go"
```

File: notification/notification.go [hunk 34]
```diff
@@ -4,7 +4,7 @@ import (
 	"fmt"
 
 	"github.com/appscode/go/log"
-	"github.com/nats-io/stan.go"
+	stan "github.com/nats-io/stan.go"
```
Comment (tahsinrahman) [inline notification/notification.go]:
```suggestion
	"github.com/nats-io/stan.go"
```

**Comment by tamalsaha** [comment]
We should be usingunio-logr here. 

`ulogr.NewLogger(klog, nats-logr)`

**Comment by tahsinrahman** [comment]
> We should be using union-logr here.
> 
> `ulogr.NewLogger(klog, nats-logr)`
> 
> And a flag `--logToNats` which is default false. So, ulogr will be just klog.

We should use ulogr only where pharmer is running as nats client mode.
Only change is needed in`apiserver/` directory, we set ulogr, then we’ll get all logs in nats, no additional change is needed

For all other cases, where pharmer running as machine controller pod and cli tool(cmd/ dir), using klogr should be ok

File: apiserver/create.go [hunk 35]
```diff
@@ -2,13 +2,17 @@ package apiserver
...
+	"github.com/go-logr/logr"
+	stan "github.com/nats-io/stan.go"
 	api "github.com/pharmer/pharmer/apis/v1beta1"
 	"github.com/pharmer/pharmer/apiserver/options"
 	"github.com/pharmer/pharmer/cloud"
 	"github.com/pharmer/pharmer/store"
+	natslogr "gomodules.xyz/nats-logr"
+	ulogr "gomodules.xyz/union-logr"
 	"k8s.io/klog/klogr"
 )
```
Comment (tahsinrahman) [inline apiserver/create.go]: move to util.go

File: apiserver/nats.go (hunk 15 above)
Comment (tahsinrahman) [inline apiserver/nats.go]:
```suggestion
	"github.com/nats-io/stan.go"
```

File: cmds/server.go [hunk 36]
```diff
@@ -42,34 +45,47 @@ func newCmdServer() *cobra.Command {
...
 //const ClientID = "worker-x"
 
-func runServer(storeProvider store.Interface, conn stan.Conn) error {
+func runServer(storeProvider store.Interface, conn stan.Conn, natsurl string, logToNats bool) error {
 
 	//defer apiserver.LogCloser(conn)
 
 	server := apiserver.New(conn)
-	err := server.CreateCluster(storeProvider)
+	err := server.CreateCluster(storeProvider, natsurl, logToNats)
```
Comment (tamalsaha) [inline cmds/server.go]: Can we just pass the logger here?

File: apiserver/util.go [hunk 37]
```diff
@@ -0,0 +1,64 @@
...
+
+	scope := cloud.NewScope(cloud.NewScopeParams{
+		Cluster:       cluster,
+		StoreProvider: storeProvider.Owner(obj.UserID),
+	})
+
+	return obj, scope, nil
+}
+
+func newLogger(operation *api.Operation, scope *cloud.Scope, natsurl string, logToNats bool) logr.Logger {
```
Comment (tahsinrahman) [inline apiserver/util.go]: Just pass clustername instead if full scope object

File: cmds/server.go (hunk 36 above)
Comment (tahsinrahman) [reply cmds/server.go]: No, we don’t know operation object here, that is used to create nats-logr, besides operation object will be different for every request

**Pull request merged**